<a href="https://colab.research.google.com/github/sofiapfm21/Python_Bootcamp/blob/assignment-1/day_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Day 1 Exercises
## Data Types

#### 1. What types are involved in the following expressions? What data type will each expression evaluate to?


In [1]:
8 * 2.5

20.0

his is an arithmetic operator for multiplication! After running it, the answer is 20.00!

In [2]:
9 / 2

4.5


This is another operator but for division instead because of the slash. After running it, the answer is 4.5

In [3]:
1.5 * 2 >= 7 - 3

False

So this would be an example of a boolean operator which is another form of data type. It has only 2 values, true or false. After running it, the answer is false. We are comparing 1.5 being multplied by 2 AND if its greather than or equal to 7-3. Basically we are comparing the two expressions, the multiplication and then subtraction.

#### 2. In which order will the expressions be evaluated?

In [4]:
6 * 3 + 7 * 4

46

So based off the lecture slide of order of operation, we learned that python follows a specific order when evaluating expressions. for the sequence above, the numbers with the star signs would be evaluated first because because theyre being multiplied first and then after it would be both answers added together. so first its 6 times 3 which is 18. and then 7 times 4 which is 28. add both answers together 18 & 28 and the answer is 46!

In [5]:
5 - 2 * 3 ** 4

-157

So the first order would be the 2 stars sign which is an operator for exponentiation. then it would be multiplied by 2. and then after it would be subtracted by 5. and the answer is -157.

In [6]:
(5 - 2) * 3 ** 4

243

the numbers being subtracted in parenthes would go first and then 3**4 would go second and both answers would be multplied last.

In [7]:
5 + 2 >= 3 * 4

False

3* 4 would go first and then 5+2 would second and then both answers would be compared as greater than or equal to.

## Comments and Errors

#### 1. Which of these expressions results in an error? Why does each error occur?

In [8]:
((((5 * 4 ** 7))))

81920

This does not result in an error! The answer is 81920 after running it.

In [9]:
84 *= 0.5 / 7

SyntaxError: 'literal' is an illegal expression for augmented assignment (868797654.py, line 1)

After running this, it comes out as an error. This is because it shows up as 'literal' being an illegal expression. Since this sequence is reffered to as an augmented assigmemnt operator, whatever is on the left side has to be a variable which means not a constant number like 84! So it can be changed as a word to either num or something else. and if we still want 84, we would have to define the variable with the value of 84. so like num=84 and then we could run it again.

In [10]:
(-(-(-(-5 * (4 + 3)))))

35

This has no error, it results in an answer of 35.

In [ ]:
5 * 3 = weight

This results in an error because again, the left side cant be a number by itself '5'. it has to be the name of a variable. so it can be reversed instead, weight=5*3

In [ ]:
73 / -----------5

This doesnt result in an error, the answer is -14.6

#### 2. Write a block of code to check if the value is a string. If it is, print out a message saying it is valid. Otherwise, raise an exception with a message saying the value must be a string.

In [ ]:
value = 6

# Your code goes here

I dont think this was covered in the class... im confused.


#### 3. What are the following errors?

In [ ]:
arr = [1, 2, 3]
print(arr[3])

The error shows up as an index error because the sequence is trying to access an element in the list at an index that doesnt exist. We learned that python lists start at 0 instead of counting from 1. so the right response would be sqaure bracket, 0, 1 and 2 sqaure bracket. and then it would be arr2 instead of arr3.

In [11]:
fullName = "Taylor Swift"
print(full_name)

NameError: name 'full_name' is not defined

we learned that for any variable, we need to define it and it has to be written properly. so it would actually have to be full_name= "Taylor Swift". Once it is defined, the command can run properly. I added the proper command above to fix the error!

In [12]:
2 + '2' + 4

TypeError: unsupported operand type(s) for +: 'int' and 'str'

So this results in an error because its trying to mix two formats together (numbers and text) which python doesnt allow!! So we would need to update and ensure that the elements are the same type, by converting the string '2' to an integer. this would ensure that all of the values are numbers and can be added properly. i will show the updated code above.

## Functions

#### 1. Following the function design recipe, define a function that converts kilometers into miles. (Assume there are 1.6 kilometers in a mile.)

In [13]:
# Your code goes here

def convert_km_to_miles(km):
  """ Convert kilometers to miles. Assume 1.6 kilometers in a mile."""
  miles= km/1.6
  return miles

In [14]:
convert_km_to_miles(5)

3.125

#### 2. What value is printed in the below code?

In [15]:
answer = 3

def answer_to_everything():
    '''Return the answer to life, the universe, and everything'''
    answer = 42
    return answer

answer_to_everything()

print(answer)

3


it gave me 3. this is because its written as answer= 3.


#### 3. Complete the body of the function. You can assume num is always >= 0.

In [ ]:
def repeat(string, num):
    ''' Return string repeated num times.
    >>> repeat('yes', 4)
    'yesyesyesyes'
    >>> repeat('no', 0)
    # your expected output here
    >>> repeat('yesnomaybe', 3)
    # your expected output here
    '''

    return (string * num)

In [ ]:
repeat ("hi", 3)

'hihihi'

In [ ]:
repeat ("hello", 4)

'hellohellohellohello'

#### 4. Complete the function below. You can assume that number will always be a string with 10 digits and that the first 3 digits will always be the area code.

In [ ]:
def is_416_number(number):
    ''' Check whether the number has a 416 area code.
    >>> is_416_number('(416)-555-5555')
    True
    >>> is_416_number('514 416 5555')
    False
    >>> is_416_number('4165554160')
    True
    '''
    # Your code goes here
    # Hint: replace () and spaces with nothing
    clean =number.replace("(","").replace(")","")
    return clean [0:3]== "416"


In [ ]:
is_416_number('4168880888')

True

In [ ]:
is_416_number('6474579587')

False